src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -529,6 +568,77 @@ private CachingAmazonDynamoDbStreams(AmazonDynamoDBStreams amazonDynamoDbStreams
         GuavaCacheMetrics.monitor(meterRegistry, iteratorCache, cn + ".GetShardIterator");
     }
 
+    /**
+     * Gets the {@code DescribeStreamResult} from the DescribeStream API.
+     * @param describeStreamRequest Describe stream request.
+     *
+     * @return Stream details for the given request with all the shards currently available.
+     *      Throws exceptions that AmazonDynamoDBStreams describeStream could potentially throw.
+     * @throws ResourceNotFoundException
+     *         The requested resource could not be found. The stream might not be specified correctly.
+     * @throws LimitExceededException
+     *         The requested resource exceeds the maximum number allowed, or the number of concurrent stream requests
+     *         exceeds the maximum number allowed (5).
+     */
+    protected DescribeStreamResult loadStreamDescriptionForAllShards(DescribeStreamRequest describeStreamRequest)
+        throws AmazonDynamoDBException, ResourceNotFoundException {
+
+        List<Shard> allShards = new ArrayList<>();
+        String streamArn = describeStreamRequest.getStreamArn();
+        String lastShardId = describeStreamRequest.getExclusiveStartShardId();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("cache miss in describe stream cache for key: {}", streamArn);
+        }
+        // TODO metrics: count cache miss
+
+        DescribeStreamResult result;
+        do {
+            describeStreamRequest = new DescribeStreamRequest().withStreamArn(streamArn)
+                .withExclusiveStartShardId(lastShardId);
+            result = super.describeStream(describeStreamRequest);
+
+            if (result.getStreamDescription().getShards() != null
+                && !result.getStreamDescription().getShards().isEmpty()) {
+                List<Shard> shards = result.getStreamDescription().getShards();
+                allShards.addAll(shards);
+                lastShardId = result.getStreamDescription().getLastEvaluatedShardId();
+            } else { // If the shards list is null/empty exit since there shouldn't be more shards.
+                lastShardId = null;
+                result.getStreamDescription().setLastEvaluatedShardId(null);
+            }
```

Comment:
I am not sure this assumption is correct. The docs say: “If LastEvaluatedShardId is not empty, it does not necessarily mean that there is more data in the result set. The only way to know when you have reached the end of the result set is when LastEvaluatedShardId is empty.“

So I think you need to keep going until you get a page that has an empty LastEvaluatedShardId regardless of whether there are shards in the list.

Reply:
Good catch.   Updated it to stop processing if LastEvaluatedShardId is null or empty.
impl and tests updated a032f5ee47ba57b46c150adadbf3552a1458a2af and 4114efc6a9911363f653a86e60abe9d0175f22c4